### PR TITLE
FSEventStream: fix crash when asking for extended data; support dispatch queues (#12007)

### PR DESCRIPTION
**Pull request by abock**
Implements support for `FSEventStreamCreateFlags.UseExtendedData`, fixing #12007.

When `.UseExtendedData` is specified, the event data type changes from `CFString` to `CFDictionary`; the dictionary contains the path (`path` key) and inode (`fileID` key) information for the file and may be extended in the future with other fields. Previously this was crashing because we assumed `CFString` always.

Further add a convenience constructor for monitoring a single path, add the missing `UnscheduleFromRunLoop` APIs, and add `SetDispatchQueue` to allow using dispatch queues directly instead of run loops.

Finally, this PR adds a fairly exhaustive file system test which covers the existing (non-extended) and fixed (extended) creation modes, along with using a dispatch queue instead of run loop.

Fixes https://github.com/xamarin/xamarin-macios/issues/12007

### CODE REVIEW

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/CoreServices/FSEvents.cs [hunk 1]
```diff
@@ -259,12 +288,37 @@ static void FreeGCHandle (IntPtr gchandle)
...
+					var fileIdHandle = CFDictionaryGetValue (
+						eventDataHandle,
+						kFSEventStreamEventExtendedFileIDKey.Handle);
+					if (fileIdHandle != IntPtr.Zero)
+						CFNumberGetValue (fileIdHandle, 4 /*kCFNumberSInt64Type*/, out fileId);
+				}
+
+				events[i] = new FSEvent
+				{
+					Id = (ulong)Marshal.ReadInt64 (eventIds, i * 8),
```
Comment (Therzok) [inline src/CoreServices/FSEvents.cs:317]: This looks like something that could be wrapped into a `ReadOnlySpan<ulong>()`. I think it's been done already in a similar fashion over at dotnet/runtime in FileSystemWatcher.OSX.cs

File: src/CoreServices/FSEvents.cs [hunk 2]
```diff
@@ -248,6 +254,29 @@ static void FreeGCHandle (IntPtr gchandle)
 			GCHandle.FromIntPtr (gchandle).Free ();
 		}
 
+		[DllImport (Constants.CoreFoundationLibrary)]
+		static extern IntPtr CFArrayGetValueAtIndex (IntPtr array, int index);
```
[the remark this answers is not in the document]
Comment (abock) [reply src/CoreServices/FSEvents.cs]: Yeah, I forgot to backpedal the duplicate P/Invokes when integrating to the build. The issue here is that it takes hours to provision a machine with the dependencies to actually build all of Xamarin.Mac, and a lot of the low level CF APIs are internal. I added extended data support out-of-band to iterate quickly on it. Really wish all the CF P/Invokes were cleanly organized and public 😄

File: tests/monotouch-test/CoreServices/FSEventStreamTest.cs [hunk 3]
```diff
@@ -5,24 +5,226 @@
...
 
 	[TestFixture]
 	[Preserve (AllMembers = true)]
-	public class FSEventStreamTest {
+	public sealed class FSEventStreamTest {
+		const string TestUid = "com.microsoft.xamarin-macios.fs-events-tests";
 
 		[Test]
-		public void Create ()
+		public void TestFileEvents ()
```
[the remark this answers is not in the document]
Comment (abock) [reply tests/monotouch-test/CoreServices/FSEventStreamTest.cs:29]: I backed the tests off to just create 256 files on disk regardless of rlmits. Bot IO is too slow and the test times out. The important part is that the test creates files, observes them through FSEventsStream, deletes files, and observes those deletions as well, for both extended-data and non-extended-data paths.

File: src/CoreServices/FSEvents.cs (hunk 1 above)
Comment (abock) [reply src/CoreServices/FSEvents.cs:300]: Sure, but this code was just moved from a few lines above. This isn't new code, so I didn't go around changing things.

Comment (abock) [reply src/CoreServices/FSEvents.cs:300]: Er, no, that would mean the underlying native alloc'd memory would have to stick around. It's releated once the callback returns.

**Comment by abock** [comment]
I forgot to mention, the reason I added the `SetDispatchQueue` API _in this PR_ is to enable the new tests. The test needs to block the main thread for the sake of NUnit, which means we can't easily use the main run loop for dispatching FS events, and it's far simpler _and more appropriate_ to create a private dispatch queue than it is a non-main-run-loop. I plan on an additional PR to bring the binding up to parity - there are missing `dev_t` based APIs that were added at the time extended data events were added (which this PR fixes) that are necessary for root-path-agnostic file watches. Depending on the velocity of this PR though I might tack the additions on to this PR as additional commits. YOLO. ❤️

**Comment by abock** [comment]
@rolfbjarne I have another patch incoming for the two missing `dev_t` based APIs as well, which become more useful with extended data events (path agnostic device ID and inode-based watching). Would you prefer I layer it into this PR or post a subsequent PR?

**Comment by rolfbjarne** [comment]
@abock new PR is best I think, doing it in chunks usually ends up easier :)

**Comment by abock** [comment]
@rolfbjarne okay, looking at the `dev_t` stuff, it's going to introduce some combinatorial .ctor explosion, and there's a chance to clean up the `sinceWhenId` parameter as well to make `FSEventStream` creation cleaner and more fluent. With that coming work in mind, I added commit 5086857 to drop the new `string pathToWatch` .ctor override to ensure that convenience API addition isn't released and frozen in time (yet). I can rebase if you'd like but figured a separate commit would be easier review. Apologies for the noise :-/

**Review by dalexsoto — APPROVED** [review]
nice ❤️

**Pull request merged**